outlines/models/exllamav2.py

```diff
@@ -1,121 +1,217 @@
+            for stop_at in generation_parameters.stop_at:
+                stop_conditions.append(stop_at)
+        exllamav2_params["stop_conditions"] = stop_conditions
+        exllamav2_params["seed"] = seed
+
+        gen_settings = ExLlamaV2Sampler.Settings()
+        if sampling_parameters.temperature is not None:
+            gen_settings.temperature = sampling_parameters.temperature
+        if sampling_parameters.top_p is not None:
+            gen_settings.top_p = sampling_parameters.top_p
+        if sampling_parameters.top_k is not None:
+            gen_settings.top_k = sampling_parameters.top_k
+        gen_settings.logits_processor = structure_logits_processor
+        exllamav2_params["gen_settings"] = gen_settings
+        if sampling_parameters.num_samples > 1:
+            prompts = prompts * sampling_parameters.num_samples
+            exllamav2_params["max_new_tokens"] = (
+                exllamav2_params["max_new_tokens"] * sampling_parameters.num_samples
+            )
+
+        if len(prompts) == 1:
+            prompts = prompts[0]
 
-        self.past_seq = seq_tensor
+        return exllamav2_params, prompts
 
-        return self.model.forward(
-            seq_tensor[-1:].view(1, -1), self.cache, loras=[self.lora]
+    def reformat_output(
+        self, output: Union[str, List[str]], sampling_parameters: SamplingParameters
+    ):
+        if isinstance(output, str):
+            return output
+        if len(output) == 1:
+            return output[0]
+        if sampling_parameters.num_samples > 1:
+            if len(output) == sampling_parameters.num_samples:
+                return output
+            assert len(output) % sampling_parameters.num_samples == 0
+            num_items_per_sample = len(output) // sampling_parameters.num_samples
+            new_output = []
+            for i in range(sampling_parameters.num_samples):
+                curr_sample = []
+                for j in range(num_items_per_sample):
+                    curr_sample.append(output[i * num_items_per_sample + j])
+                new_output.append(curr_sample)
+            return new_output
+        return output
+
+    def generate(
+        self,
+        prompts: Union[str, List[str]],
+        generation_parameters: GenerationParameters,
+        structure_logits_processor,
+        sampling_parameters: SamplingParameters,
+        **exllamav2_params: Unpack[ExllamaV2Params],
+    ) -> Union[str, List[str]]:
+        exllamav2_params, prompts = self.prepare_generation_parameters(
+            prompts,
+            generation_parameters,
+            sampling_parameters,
+            structure_logits_processor,
+        )
+        output = self.generator.generate(
+            prompt=prompts,
+            gen_settings=exllamav2_params["gen_settings"],
+            max_new_tokens=min(exllamav2_params["max_new_tokens"]),
+            completion_only=True,
+            encode_special_tokens=False,
```

Comment:
Why don't we encode special tokens?

Reply:
I mainly did this because in exllamav2 the default value for encode_special_tokens was False. But I can't think of much reason for it needing to be False. Let me try changing to True and then running the tests